{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "import argparse\n",
    "import copy\n",
    "from gtfparse import read_gtf\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "from fuc import pybed\n",
    "\n",
    "from Bio import AlignIO\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "import re"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Create BED file for all DE isoforms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# isoforms_df = read_gtf(\"../data/isoform_analysis/mm10.aligned.sorted_corrected.gtf\")\n",
    "isoforms_df = pd.read_csv(\"../data/isoform_analysis/mm10.aligned.sorted_corrected.gtf\", sep=\"\\t\", header=None)\n",
    "isoforms_df[\"transcript_id\"] = isoforms_df[8].str.split(\" \", expand=True)[1].str.split(\"\\\"\", expand=True)[1]\n",
    "isoforms_df.columns = [\"seqname\", \"insname\", \"feature\", \"start\", \"end\", \"score\", \"strand\", \"score1\", \"info\", \"transcript_id\"]\n",
    "isoforms_df = isoforms_df[isoforms_df[\"feature\"] == \"transcript\"]\n",
    "isoforms_df = isoforms_df[[\"seqname\", \"start\", \"end\", \"strand\", \"transcript_id\"]]\n",
    "#Reorder columns to create a bed file. Also, make 0-based coordinates\n",
    "isoforms_df[\"start\"] = isoforms_df[\"start\"] - 1\n",
    "isoforms_df[\"end\"] = isoforms_df[\"end\"] - 1\n",
    "isoforms_df.columns = [\"Chromosome\", \"Start\", \"End\", \"Strand\", \"Isoform\"]\n",
    "isoforms_df[\"Category\"] = 0\n",
    "isoforms_df = isoforms_df[[\"Chromosome\", \"Start\", \"End\", \"Isoform\", \"Category\", \"Strand\"]]\n",
    "remove_chr = isoforms_df[\"Chromosome\"].unique().tolist()[22:]\n",
    "isoforms_df = isoforms_df[~isoforms_df[\"Chromosome\"].isin(remove_chr)]\n",
    "#Only keep DE isoforms\n",
    "brf1_ctrl_inf = pd.read_csv(\"../data/DE_analysis/glaunsinger_Brf1_48hpi_v_Ctrl_48_DE_analysis_mm10.txt\", sep=\"\\s+\")\n",
    "de_transcripts = brf1_ctrl_inf[brf1_ctrl_inf[\"adj.P.Val\"] <= 0.05][\"Transcript\"].unique().tolist()\n",
    "isoforms_df = isoforms_df[isoforms_df[\"Isoform\"].isin(de_transcripts)]\n",
    "isoforms_df\n",
    "bf = pybed.BedFrame.from_frame(meta=[], data=isoforms_df)\n",
    "bf.to_file('../data/bed_files/coords_for_allDE_isoforms_ctrl48_brf148.bed')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Create BED file for DE exons in isoforms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Isoform</th>\n",
       "      <th>Category</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>50323</th>\n",
       "      <td>chr1</td>\n",
       "      <td>55003174</td>\n",
       "      <td>55003204</td>\n",
       "      <td>PB.316.57</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50324</th>\n",
       "      <td>chr1</td>\n",
       "      <td>55003292</td>\n",
       "      <td>55003489</td>\n",
       "      <td>PB.316.57</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50325</th>\n",
       "      <td>chr1</td>\n",
       "      <td>55005801</td>\n",
       "      <td>55005922</td>\n",
       "      <td>PB.316.57</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50326</th>\n",
       "      <td>chr1</td>\n",
       "      <td>55006170</td>\n",
       "      <td>55006382</td>\n",
       "      <td>PB.316.57</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50327</th>\n",
       "      <td>chr1</td>\n",
       "      <td>55007479</td>\n",
       "      <td>55007716</td>\n",
       "      <td>PB.316.57</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4451932</th>\n",
       "      <td>chrY</td>\n",
       "      <td>1023301</td>\n",
       "      <td>1023473</td>\n",
       "      <td>PB.23977.40</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4451933</th>\n",
       "      <td>chrY</td>\n",
       "      <td>1028187</td>\n",
       "      <td>1028590</td>\n",
       "      <td>PB.23977.40</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4455849</th>\n",
       "      <td>chrY</td>\n",
       "      <td>75606590</td>\n",
       "      <td>75609220</td>\n",
       "      <td>PB.24050.1</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4455876</th>\n",
       "      <td>chrY</td>\n",
       "      <td>83165948</td>\n",
       "      <td>83167364</td>\n",
       "      <td>PB.24063.1</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4455886</th>\n",
       "      <td>chrY</td>\n",
       "      <td>87029086</td>\n",
       "      <td>87030349</td>\n",
       "      <td>PB.24068.1</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2900 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chromosome     Start       End      Isoform  Category Strand\n",
       "50323         chr1  55003174  55003204    PB.316.57         0      -\n",
       "50324         chr1  55003292  55003489    PB.316.57         0      -\n",
       "50325         chr1  55005801  55005922    PB.316.57         0      -\n",
       "50326         chr1  55006170  55006382    PB.316.57         0      -\n",
       "50327         chr1  55007479  55007716    PB.316.57         0      -\n",
       "...            ...       ...       ...          ...       ...    ...\n",
       "4451932       chrY   1023301   1023473  PB.23977.40         0      +\n",
       "4451933       chrY   1028187   1028590  PB.23977.40         0      +\n",
       "4455849       chrY  75606590  75609220   PB.24050.1         0      +\n",
       "4455876       chrY  83165948  83167364   PB.24063.1         0      +\n",
       "4455886       chrY  87029086  87030349   PB.24068.1         0      +\n",
       "\n",
       "[2900 rows x 6 columns]"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# isoforms_df = read_gtf(\"../data/isoform_analysis/mm10.aligned.sorted_corrected.gtf\")\n",
    "isoforms_df = pd.read_csv(\"../data/isoform_analysis/mm10.aligned.sorted_corrected.gtf\", sep=\"\\t\", header=None)\n",
    "isoforms_df[\"transcript_id\"] = isoforms_df[8].str.split(\" \", expand=True)[1].str.split(\"\\\"\", expand=True)[1]\n",
    "isoforms_df.columns = [\"seqname\", \"insname\", \"feature\", \"start\", \"end\", \"score\", \"strand\", \"score1\", \"info\", \"transcript_id\"]\n",
    "isoforms_df = isoforms_df[isoforms_df[\"feature\"] == \"exon\"]\n",
    "isoforms_df = isoforms_df[[\"seqname\", \"start\", \"end\", \"strand\", \"transcript_id\"]]\n",
    "#Reorder columns to create a bed file. Also, make 0-based coordinates\n",
    "isoforms_df[\"start\"] = isoforms_df[\"start\"] - 1\n",
    "isoforms_df[\"end\"] = isoforms_df[\"end\"] - 1\n",
    "isoforms_df.columns = [\"Chromosome\", \"Start\", \"End\", \"Strand\", \"Isoform\"]\n",
    "isoforms_df[\"Category\"] = 0\n",
    "isoforms_df = isoforms_df[[\"Chromosome\", \"Start\", \"End\", \"Isoform\", \"Category\", \"Strand\"]]\n",
    "remove_chr = isoforms_df[\"Chromosome\"].unique().tolist()[22:]\n",
    "isoforms_df = isoforms_df[~isoforms_df[\"Chromosome\"].isin(remove_chr)]\n",
    "#Only keep DE isoforms\n",
    "brf1_ctrl_inf = pd.read_csv(\"../data/DE_analysis/glaunsinger_Brf1_48hpi_v_Ctrl_48_DE_analysis_mm10.txt\", sep=\"\\s+\")\n",
    "de_transcripts = brf1_ctrl_inf[brf1_ctrl_inf[\"adj.P.Val\"] <= 0.05][\"Transcript\"].unique().tolist()\n",
    "isoforms_df = isoforms_df[isoforms_df[\"Isoform\"].isin(de_transcripts)]\n",
    "isoforms_df\n",
    "# bf = pybed.BedFrame.from_frame(meta=[], data=isoforms_df)\n",
    "# bf.to_file('../data/bed_files/exon_coords_for_DE_isoforms_ctrl48_brf148.bed')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "bf = pybed.BedFrame.from_frame(meta=[], data=isoforms_df)\n",
    "bf.to_file('../data/bed_files/exon_coords_for_DE_isoforms_ctrl48_brf148.bed')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "# brf1_ctrl_inf = pd.read_csv(\"../data/DE_analysis/glaunsinger_Brf1_48hpi_v_Ctrl_48_DE_analysis_mm10.txt\", sep=\"\\s+\")\n",
    "# de_transcripts = brf1_ctrl_inf.loc[(brf1_ctrl_inf[\"adj.P.Val\"] <= 0.05) & ((brf1_ctrl_inf[\"logFC\"] >= 1.5))][\"Transcript\"].unique().tolist()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Create BED file for all introns in isoforms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Junction</th>\n",
       "      <th>Category</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>41844</th>\n",
       "      <td>chr1</td>\n",
       "      <td>55003205</td>\n",
       "      <td>55003291</td>\n",
       "      <td>PB.316.57_junction_1</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41845</th>\n",
       "      <td>chr1</td>\n",
       "      <td>55003490</td>\n",
       "      <td>55005800</td>\n",
       "      <td>PB.316.57_junction_2</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41846</th>\n",
       "      <td>chr1</td>\n",
       "      <td>55005923</td>\n",
       "      <td>55006169</td>\n",
       "      <td>PB.316.57_junction_3</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41847</th>\n",
       "      <td>chr1</td>\n",
       "      <td>55006383</td>\n",
       "      <td>55007478</td>\n",
       "      <td>PB.316.57_junction_4</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41848</th>\n",
       "      <td>chr1</td>\n",
       "      <td>55007717</td>\n",
       "      <td>55012100</td>\n",
       "      <td>PB.316.57_junction_5</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3689165</th>\n",
       "      <td>chrY</td>\n",
       "      <td>1014792</td>\n",
       "      <td>1016055</td>\n",
       "      <td>PB.23977.40_junction_6</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3689166</th>\n",
       "      <td>chrY</td>\n",
       "      <td>1016684</td>\n",
       "      <td>1017362</td>\n",
       "      <td>PB.23977.40_junction_7</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3689167</th>\n",
       "      <td>chrY</td>\n",
       "      <td>1017508</td>\n",
       "      <td>1019984</td>\n",
       "      <td>PB.23977.40_junction_8</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3689168</th>\n",
       "      <td>chrY</td>\n",
       "      <td>1020155</td>\n",
       "      <td>1023300</td>\n",
       "      <td>PB.23977.40_junction_9</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3689169</th>\n",
       "      <td>chrY</td>\n",
       "      <td>1023474</td>\n",
       "      <td>1028186</td>\n",
       "      <td>PB.23977.40_junction_10</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2427 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chromosome     Start       End                 Junction  Category  \\\n",
       "41844         chr1  55003205  55003291     PB.316.57_junction_1         0   \n",
       "41845         chr1  55003490  55005800     PB.316.57_junction_2         0   \n",
       "41846         chr1  55005923  55006169     PB.316.57_junction_3         0   \n",
       "41847         chr1  55006383  55007478     PB.316.57_junction_4         0   \n",
       "41848         chr1  55007717  55012100     PB.316.57_junction_5         0   \n",
       "...            ...       ...       ...                      ...       ...   \n",
       "3689165       chrY   1014792   1016055   PB.23977.40_junction_6         0   \n",
       "3689166       chrY   1016684   1017362   PB.23977.40_junction_7         0   \n",
       "3689167       chrY   1017508   1019984   PB.23977.40_junction_8         0   \n",
       "3689168       chrY   1020155   1023300   PB.23977.40_junction_9         0   \n",
       "3689169       chrY   1023474   1028186  PB.23977.40_junction_10         0   \n",
       "\n",
       "        Strand  \n",
       "41844        -  \n",
       "41845        -  \n",
       "41846        -  \n",
       "41847        -  \n",
       "41848        -  \n",
       "...        ...  \n",
       "3689165      +  \n",
       "3689166      +  \n",
       "3689167      +  \n",
       "3689168      +  \n",
       "3689169      +  \n",
       "\n",
       "[2427 rows x 6 columns]"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "junctions_df = pd.read_csv(\"../data/isoform_analysis/mm10.aligned.sorted_junctions.txt\", sep=\"\\s+\")\n",
    "junctions_df = junctions_df[[\"isoform\", \"chrom\", \"strand\", \"junction_number\", \"genomic_start_coord\", \n",
    "                             \"genomic_end_coord\"]]\n",
    "junctions_df[\"Junction\"] = junctions_df[\"isoform\"] + \"_\" + junctions_df[\"junction_number\"]\n",
    "#Only keep DE isoforms\n",
    "brf1_ctrl_inf = pd.read_csv(\"../data/DE_analysis/glaunsinger_Brf1_48hpi_v_Ctrl_48_DE_analysis_mm10.txt\", sep=\"\\s+\")\n",
    "de_transcripts = brf1_ctrl_inf[brf1_ctrl_inf[\"adj.P.Val\"] <= 0.05][\"Transcript\"].unique().tolist()\n",
    "junctions_df = junctions_df[junctions_df[\"isoform\"].isin(de_transcripts)]\n",
    "\n",
    "junctions_df = junctions_df.drop(columns=[\"isoform\", \"junction_number\"])\n",
    "junctions_df.columns = [\"Chromosome\", \"Strand\", \"Start\", \"End\", \"Junction\"]\n",
    "junctions_df[\"Start\"] = junctions_df[\"Start\"] - 1\n",
    "junctions_df[\"End\"] = junctions_df[\"End\"] - 1\n",
    "junctions_df[\"Category\"] = 0\n",
    "junctions_df = junctions_df[[\"Chromosome\", \"Start\", \"End\", \"Junction\", \"Category\", \"Strand\"]]\n",
    "junctions_df = junctions_df[~junctions_df[\"Chromosome\"].isin(remove_chr)]\n",
    "junctions_df\n",
    "# bf = pybed.BedFrame.from_frame(meta=[], data=junctions_df)\n",
    "# bf.to_file('../data/bed_files/intron_coords_for_DE_isoforms_ctrl48_brf148.bed')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "bf = pybed.BedFrame.from_frame(meta=[], data=junctions_df)\n",
    "bf.to_file('../data/bed_files/intron_coords_for_DE_isoforms_ctrl48_brf148.bed')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# -------------------------Extract intact AB box B2 SINEs--------------------------"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3063370</td>\n",
       "      <td>3063529</td>\n",
       "      <td>active_b2</td>\n",
       "      <td>169.812951</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>chr1</td>\n",
       "      <td>4069693</td>\n",
       "      <td>4069850</td>\n",
       "      <td>active_b2</td>\n",
       "      <td>6.092839</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>chr1</td>\n",
       "      <td>4173454</td>\n",
       "      <td>4173632</td>\n",
       "      <td>active_b2</td>\n",
       "      <td>8.067949</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>chr1</td>\n",
       "      <td>4200994</td>\n",
       "      <td>4201177</td>\n",
       "      <td>active_b2</td>\n",
       "      <td>15.968930</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>chr1</td>\n",
       "      <td>4202363</td>\n",
       "      <td>4202566</td>\n",
       "      <td>active_b2</td>\n",
       "      <td>8.169802</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>366855</th>\n",
       "      <td>chrX</td>\n",
       "      <td>168649121</td>\n",
       "      <td>168649297</td>\n",
       "      <td>active_b2</td>\n",
       "      <td>7.776876</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>366862</th>\n",
       "      <td>chrX</td>\n",
       "      <td>168735132</td>\n",
       "      <td>168735291</td>\n",
       "      <td>active_b2</td>\n",
       "      <td>12.550832</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>366868</th>\n",
       "      <td>chrX</td>\n",
       "      <td>168788247</td>\n",
       "      <td>168788437</td>\n",
       "      <td>active_b2</td>\n",
       "      <td>14.658536</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>366948</th>\n",
       "      <td>chrX</td>\n",
       "      <td>169867459</td>\n",
       "      <td>169867618</td>\n",
       "      <td>active_b2</td>\n",
       "      <td>15.458246</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>367025</th>\n",
       "      <td>chrY</td>\n",
       "      <td>862023</td>\n",
       "      <td>862211</td>\n",
       "      <td>active_b2</td>\n",
       "      <td>7.905422</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>45403 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Chromosome      Start        End       Name       Score Strand\n",
       "2            chr1    3063370    3063529  active_b2  169.812951      -\n",
       "46           chr1    4069693    4069850  active_b2    6.092839      -\n",
       "49           chr1    4173454    4173632  active_b2    8.067949      -\n",
       "51           chr1    4200994    4201177  active_b2   15.968930      +\n",
       "53           chr1    4202363    4202566  active_b2    8.169802      +\n",
       "...           ...        ...        ...        ...         ...    ...\n",
       "366855       chrX  168649121  168649297  active_b2    7.776876      +\n",
       "366862       chrX  168735132  168735291  active_b2   12.550832      +\n",
       "366868       chrX  168788247  168788437  active_b2   14.658536      +\n",
       "366948       chrX  169867459  169867618  active_b2   15.458246      -\n",
       "367025       chrY     862023     862211  active_b2    7.905422      -\n",
       "\n",
       "[45403 rows x 6 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "active_b2_df1 = pd.read_csv(\"../data/isoform_analysis/GSM2274627_BGJK003_all_B2_SINES_rpkm.txt\", sep='\\s+')\n",
    "active_b2_df2 = pd.read_csv(\"../data/isoform_analysis/GSM2274628_BGJK005_all_B2_SINES_rpkm.txt\", sep='\\s+')\n",
    "active_b2 = active_b2_df1.merge(active_b2_df2, on=[\"chrom\", \"start\", \"end\", \"strand\", \"length\"])\n",
    "active_b2[\"Average_RPKM\"] = active_b2[[\"RPKM_x\", \"RPKM_y\"]].mean(axis=1)\n",
    "active_b2 = active_b2[[\"chrom\", \"start\", \"end\", \"strand\", \"Average_RPKM\"]]\n",
    "active_b2 = active_b2[active_b2[\"Average_RPKM\"] >= 5]\n",
    "active_b2[\"Name\"] = \"active_b2\"\n",
    "active_b2 = active_b2.rename(columns={\"chrom\": \"Chromosome\", \"start\": \"Start\", \"end\": \"End\", \"strand\": \"Strand\",\n",
    "                                     \"Average_RPKM\": \"Score\"})\n",
    "active_b2 = active_b2[[\"Chromosome\", \"Start\", \"End\", \"Name\", \"Score\", \"Strand\"]]\n",
    "active_b2\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Save active b2 as bed file\n",
    "bf = pybed.BedFrame.from_frame(meta=[], data=active_b2)\n",
    "bf.to_file('../data/bed_files/active_b2.bed')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "def listToString(s): \n",
    "    \n",
    "    # initialize an empty string\n",
    "    str1 = \"\" \n",
    "    \n",
    "    # traverse the string  \n",
    "    for ele in s: \n",
    "        str1 += ele  \n",
    "    \n",
    "    # return string  \n",
    "    return str1 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_AB_coord(df, idx, genome_seq):\n",
    "    \n",
    "    if df[\"strand\"].iloc[idx] == \"+\":\n",
    "        b2_seq = listToString(list(genome_seq[df[\"START\"].iloc[idx]:df[\"STOP\"].iloc[idx]]))\n",
    "    else:\n",
    "        b2_seq = listToString(list(genome_seq[df[\"START\"].iloc[idx]:df[\"STOP\"].iloc[idx]].reverse_complement()))\n",
    "    try:\n",
    "        A = re.search(\"T[A,G]G[C,T][A,C,G,T][A,C,G,T]A[A,G][A,C,G,T][A,C,G,T]G\", b2_seq).start()\n",
    "    except AttributeError:\n",
    "        A = np.nan\n",
    "    try:\n",
    "        B = re.search(\"[A,G]GTTC[A,G]A[A,C,G,T]TCC\", b2_seq).start()\n",
    "    except AttributeError:\n",
    "        B = np.nan\n",
    "    \n",
    "    if np.isnan(A) and np.isnan(B):\n",
    "        return \"mutated\"\n",
    "    elif A >= 0 and B >= 0:\n",
    "        return \"intact\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Read in all genome sequences\n",
    "chrom_in_df = all_b2[\"chrom\"].unique().tolist()\n",
    "chrom_dict = {}\n",
    "\n",
    "for ch in chrom_in_df:\n",
    "    chrom_dict[ch] = SeqIO.read(\"/Users/sahilshah/Documents/Rotations/Britt_Glaunsinger/archive_2021_12_03__08_19_08/mouse_genome/chromosomes/{}.fa\".format(ch), 'fasta').seq.upper()\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "AB_status = []\n",
    "\n",
    "for idx in range(len(all_b2)):\n",
    "    #print(idx)\n",
    "    chromosome = all_b2[\"chrom\"].iloc[idx]\n",
    "    genome_seq = chrom_dict[chromosome]\n",
    "    AB_coord = extract_AB_coord(all_b2, idx, genome_seq)\n",
    "    AB_status.append(AB_coord)\n",
    "    \n",
    "all_b2[\"AB_status\"] = AB_status"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>START</th>\n",
       "      <th>STOP</th>\n",
       "      <th>strand</th>\n",
       "      <th>length</th>\n",
       "      <th>count</th>\n",
       "      <th>count(CPM)</th>\n",
       "      <th>RPKM</th>\n",
       "      <th>AB_status</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3013180</td>\n",
       "      <td>3013364</td>\n",
       "      <td>-</td>\n",
       "      <td>184</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3050778</td>\n",
       "      <td>3050932</td>\n",
       "      <td>-</td>\n",
       "      <td>154</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>intact</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3063370</td>\n",
       "      <td>3063529</td>\n",
       "      <td>-</td>\n",
       "      <td>159</td>\n",
       "      <td>145</td>\n",
       "      <td>23.904963</td>\n",
       "      <td>150.34568</td>\n",
       "      <td>intact</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3114021</td>\n",
       "      <td>3114219</td>\n",
       "      <td>-</td>\n",
       "      <td>198</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>mutated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3118142</td>\n",
       "      <td>3118352</td>\n",
       "      <td>+</td>\n",
       "      <td>210</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>mutated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>372918</th>\n",
       "      <td>chrY_JH584303_random</td>\n",
       "      <td>58186</td>\n",
       "      <td>58380</td>\n",
       "      <td>+</td>\n",
       "      <td>194</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>mutated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>372919</th>\n",
       "      <td>chrY_JH584303_random</td>\n",
       "      <td>73058</td>\n",
       "      <td>73168</td>\n",
       "      <td>-</td>\n",
       "      <td>110</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>mutated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>372920</th>\n",
       "      <td>chrY_JH584303_random</td>\n",
       "      <td>85023</td>\n",
       "      <td>85191</td>\n",
       "      <td>+</td>\n",
       "      <td>168</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>372921</th>\n",
       "      <td>chrY_JH584303_random</td>\n",
       "      <td>92003</td>\n",
       "      <td>92168</td>\n",
       "      <td>-</td>\n",
       "      <td>165</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>mutated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>372922</th>\n",
       "      <td>chrY_JH584303_random</td>\n",
       "      <td>133892</td>\n",
       "      <td>134097</td>\n",
       "      <td>+</td>\n",
       "      <td>205</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>mutated</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>372923 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       chrom    START     STOP strand  length  count  \\\n",
       "0                       chr1  3013180  3013364      -     184      0   \n",
       "1                       chr1  3050778  3050932      -     154      0   \n",
       "2                       chr1  3063370  3063529      -     159    145   \n",
       "3                       chr1  3114021  3114219      -     198      0   \n",
       "4                       chr1  3118142  3118352      +     210      0   \n",
       "...                      ...      ...      ...    ...     ...    ...   \n",
       "372918  chrY_JH584303_random    58186    58380      +     194      0   \n",
       "372919  chrY_JH584303_random    73058    73168      -     110      0   \n",
       "372920  chrY_JH584303_random    85023    85191      +     168      0   \n",
       "372921  chrY_JH584303_random    92003    92168      -     165      0   \n",
       "372922  chrY_JH584303_random   133892   134097      +     205      0   \n",
       "\n",
       "        count(CPM)       RPKM AB_status  \n",
       "0         0.000000    0.00000      None  \n",
       "1         0.000000    0.00000    intact  \n",
       "2        23.904963  150.34568    intact  \n",
       "3         0.000000    0.00000   mutated  \n",
       "4         0.000000    0.00000   mutated  \n",
       "...            ...        ...       ...  \n",
       "372918    0.000000    0.00000   mutated  \n",
       "372919    0.000000    0.00000   mutated  \n",
       "372920    0.000000    0.00000      None  \n",
       "372921    0.000000    0.00000   mutated  \n",
       "372922    0.000000    0.00000   mutated  \n",
       "\n",
       "[372923 rows x 9 columns]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_b2 #NOTE: B2 with only one intact portion of the AB box is considered Neither mutated nor intact and we just throw\n",
    "#them out"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Save active B2 with intact AB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3050778</td>\n",
       "      <td>3050932</td>\n",
       "      <td>B2</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3063370</td>\n",
       "      <td>3063529</td>\n",
       "      <td>B2</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>chr1</td>\n",
       "      <td>4200127</td>\n",
       "      <td>4200286</td>\n",
       "      <td>B2</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>chr1</td>\n",
       "      <td>4478363</td>\n",
       "      <td>4478536</td>\n",
       "      <td>B2</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>chr1</td>\n",
       "      <td>4572758</td>\n",
       "      <td>4572933</td>\n",
       "      <td>B2</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>372847</th>\n",
       "      <td>chrY</td>\n",
       "      <td>90453727</td>\n",
       "      <td>90453903</td>\n",
       "      <td>B2</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>372867</th>\n",
       "      <td>chrY</td>\n",
       "      <td>90759897</td>\n",
       "      <td>90760086</td>\n",
       "      <td>B2</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>372869</th>\n",
       "      <td>chrY</td>\n",
       "      <td>90791872</td>\n",
       "      <td>90792066</td>\n",
       "      <td>B2</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>372871</th>\n",
       "      <td>chrY</td>\n",
       "      <td>90844253</td>\n",
       "      <td>90844442</td>\n",
       "      <td>B2</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>372885</th>\n",
       "      <td>chrY_JH584300_random</td>\n",
       "      <td>165258</td>\n",
       "      <td>165443</td>\n",
       "      <td>B2</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>67244 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  Chromosome     Start       End Name  Score Strand\n",
       "1                       chr1   3050778   3050932   B2      0      -\n",
       "2                       chr1   3063370   3063529   B2      0      -\n",
       "50                      chr1   4200127   4200286   B2      0      -\n",
       "68                      chr1   4478363   4478536   B2      0      -\n",
       "79                      chr1   4572758   4572933   B2      0      -\n",
       "...                      ...       ...       ...  ...    ...    ...\n",
       "372847                  chrY  90453727  90453903   B2      0      -\n",
       "372867                  chrY  90759897  90760086   B2      0      -\n",
       "372869                  chrY  90791872  90792066   B2      0      -\n",
       "372871                  chrY  90844253  90844442   B2      0      -\n",
       "372885  chrY_JH584300_random    165258    165443   B2      0      +\n",
       "\n",
       "[67244 rows x 6 columns]"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "intact_b2 = all_b2[all_b2[\"AB_status\"] == \"intact\"]\n",
    "#If I want B2 SINEs that are transcriptionally active according to SINE-seq paper, otherwise I want all B2 SINEs w/ intact AB boxes\n",
    "# intact_b2 = intact_b2[intact_b2[\"RPKM\"] >= 5]\n",
    "intact_b2 = intact_b2.drop(columns=[\"length\", \"count\", \"count(CPM)\", \"RPKM\", \"AB_status\"])\n",
    "intact_b2[\"Score\"] = 0\n",
    "intact_b2[\"Name\"] = \"B2\"\n",
    "# intact_b2[\"START\"] = intact_b2[\"START\"] - 1\n",
    "# intact_b2[\"STOP\"] = intact_b2[\"STOP\"] - 1\n",
    "intact_b2 = intact_b2[[\"chrom\", \"START\", \"STOP\", \"Name\", \"Score\", \"strand\"]]\n",
    "intact_b2.columns = [\"Chromosome\", \"Start\", \"End\", \"Name\", \"Score\", \"Strand\"]\n",
    "intact_b2\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Intact B2\n",
    "bf = pybed.BedFrame.from_frame(meta=[], data=intact_b2)\n",
    "bf.to_file('bedtools_inputs/b2_w_intact_AB.bed')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Create B2 bed file\n",
    "all_b2 = pd.read_csv(\"bedtools_inputs/old_b2_files/all_b2_final.tsv\", sep=\"\\t\", header=None)\n",
    "all_b2 = all_b2[[5, 6, 7, 9, 10]].rename(columns={5: \"Chromosome\", 6: \"Start\", 7: \"End\", 9: \"Strand\",\n",
    "                                                 10: \"Name\"})\n",
    "all_b2[\"Score\"] = 0\n",
    "all_b2 = all_b2[[\"Chromosome\", \"Start\", \"End\", \"Name\", \"Score\", \"Strand\"]]\n",
    "bf = pybed.BedFrame.from_frame(meta=[], data=all_b2)\n",
    "bf.to_file('bedtools_inputs/all_b2_final.bed')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Extract isoform UTR coordinates"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read in file containing TSS and TTS coordinates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>CDS_Start</th>\n",
       "      <th>CDS_End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GL456212.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PB.24087.1</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>JH584299.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PB.24097.1</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>849</th>\n",
       "      <td>KK082442.1</td>\n",
       "      <td>142719.0</td>\n",
       "      <td>142354.0</td>\n",
       "      <td>PB.24142.42</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1105</th>\n",
       "      <td>KQ030494.1</td>\n",
       "      <td>93892.0</td>\n",
       "      <td>115515.0</td>\n",
       "      <td>PB.24270.2</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2911</th>\n",
       "      <td>chr1</td>\n",
       "      <td>150100224.0</td>\n",
       "      <td>150105782.0</td>\n",
       "      <td>PB.1006.28</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>383780</th>\n",
       "      <td>chrX</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PB.23959.3</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>384304</th>\n",
       "      <td>chrY</td>\n",
       "      <td>1010617.0</td>\n",
       "      <td>1016379.0</td>\n",
       "      <td>PB.23977.40</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>384588</th>\n",
       "      <td>chrY</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PB.24050.1</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>384601</th>\n",
       "      <td>chrY</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PB.24063.1</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>384606</th>\n",
       "      <td>chrY</td>\n",
       "      <td>87029310.0</td>\n",
       "      <td>87029726.0</td>\n",
       "      <td>PB.24068.1</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>477 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chromosome    CDS_Start      CDS_End         Name  Score Strand\n",
       "2       GL456212.1          NaN          NaN   PB.24087.1      0      +\n",
       "65      JH584299.1          NaN          NaN   PB.24097.1      0      +\n",
       "849     KK082442.1     142719.0     142354.0  PB.24142.42      0      -\n",
       "1105    KQ030494.1      93892.0     115515.0   PB.24270.2      0      +\n",
       "2911          chr1  150100224.0  150105782.0   PB.1006.28      0      +\n",
       "...            ...          ...          ...          ...    ...    ...\n",
       "383780        chrX          NaN          NaN   PB.23959.3      0      -\n",
       "384304        chrY    1010617.0    1016379.0  PB.23977.40      0      +\n",
       "384588        chrY          NaN          NaN   PB.24050.1      0      +\n",
       "384601        chrY          NaN          NaN   PB.24063.1      0      +\n",
       "384606        chrY   87029310.0   87029726.0   PB.24068.1      0      +\n",
       "\n",
       "[477 rows x 6 columns]"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "UTR_coords_df = pd.read_csv(\"../data/DE_analysis/mm10.aligned.sorted_classification.txt\", sep=\"\\s+\")\n",
    "UTR_coords_df = UTR_coords_df[[\"chrom\", \"CDS_genomic_start\",  \"CDS_genomic_end\", \"isoform\", \"strand\"]]\n",
    "UTR_coords_df[\"Score\"] = 0\n",
    "UTR_coords_df = UTR_coords_df[[\"chrom\", \"CDS_genomic_start\",  \"CDS_genomic_end\", \"isoform\", \"Score\", \"strand\"]]\n",
    "#Only keep DE isoforms\n",
    "brf1_ctrl_inf = pd.read_csv(\"../data/DE_analysis/glaunsinger_Brf1_48hpi_v_Ctrl_48_DE_analysis_mm10.txt\", sep=\"\\s+\")\n",
    "de_transcripts = brf1_ctrl_inf[brf1_ctrl_inf[\"adj.P.Val\"] <= 0.05][\"Transcript\"].unique().tolist()\n",
    "UTR_coords_df = UTR_coords_df[UTR_coords_df[\"isoform\"].isin(de_transcripts)]\n",
    "\n",
    "UTR_coords_df.columns = [\"Chromosome\", \"CDS_Start\", \"CDS_End\", \"Name\", \"Score\", \"Strand\"]\n",
    "UTR_coords_df\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [],
   "source": [
    "#UTR_coords_df.columns.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Transcript_Start</th>\n",
       "      <th>Transcript_End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>50322</th>\n",
       "      <td>chr1</td>\n",
       "      <td>55003175</td>\n",
       "      <td>55027395</td>\n",
       "      <td>PB.316.57</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75210</th>\n",
       "      <td>chr1</td>\n",
       "      <td>74288252</td>\n",
       "      <td>74304619</td>\n",
       "      <td>PB.468.7</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94407</th>\n",
       "      <td>chr1</td>\n",
       "      <td>84034264</td>\n",
       "      <td>84284599</td>\n",
       "      <td>PB.558.4</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126881</th>\n",
       "      <td>chr1</td>\n",
       "      <td>91053479</td>\n",
       "      <td>91115530</td>\n",
       "      <td>PB.654.62</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128768</th>\n",
       "      <td>chr1</td>\n",
       "      <td>91453766</td>\n",
       "      <td>91459301</td>\n",
       "      <td>PB.666.13</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4455885</th>\n",
       "      <td>chrY</td>\n",
       "      <td>87029087</td>\n",
       "      <td>87030350</td>\n",
       "      <td>PB.24068.1</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4456545</th>\n",
       "      <td>GL456212.1</td>\n",
       "      <td>82189</td>\n",
       "      <td>84538</td>\n",
       "      <td>PB.24087.1</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4456954</th>\n",
       "      <td>JH584299.1</td>\n",
       "      <td>935049</td>\n",
       "      <td>936476</td>\n",
       "      <td>PB.24097.1</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4462451</th>\n",
       "      <td>KK082442.1</td>\n",
       "      <td>141292</td>\n",
       "      <td>142839</td>\n",
       "      <td>PB.24142.42</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4474510</th>\n",
       "      <td>KQ030494.1</td>\n",
       "      <td>93088</td>\n",
       "      <td>115814</td>\n",
       "      <td>PB.24270.2</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>477 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         Chromosome  Transcript_Start  Transcript_End         Name  Score  \\\n",
       "50322          chr1          55003175        55027395    PB.316.57      0   \n",
       "75210          chr1          74288252        74304619     PB.468.7      0   \n",
       "94407          chr1          84034264        84284599     PB.558.4      0   \n",
       "126881         chr1          91053479        91115530    PB.654.62      0   \n",
       "128768         chr1          91453766        91459301    PB.666.13      0   \n",
       "...             ...               ...             ...          ...    ...   \n",
       "4455885        chrY          87029087        87030350   PB.24068.1      0   \n",
       "4456545  GL456212.1             82189           84538   PB.24087.1      0   \n",
       "4456954  JH584299.1            935049          936476   PB.24097.1      0   \n",
       "4462451  KK082442.1            141292          142839  PB.24142.42      0   \n",
       "4474510  KQ030494.1             93088          115814   PB.24270.2      0   \n",
       "\n",
       "        Strand  \n",
       "50322        -  \n",
       "75210        -  \n",
       "94407        -  \n",
       "126881       +  \n",
       "128768       -  \n",
       "...        ...  \n",
       "4455885      +  \n",
       "4456545      +  \n",
       "4456954      +  \n",
       "4462451      -  \n",
       "4474510      +  \n",
       "\n",
       "[477 rows x 6 columns]"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "isoforms_df = pd.read_csv(\"../data/isoform_analysis/mm10.aligned.sorted_corrected.gtf\", sep=\"\\t\", header=None)\n",
    "isoforms_df[\"transcript_id\"] = isoforms_df[8].str.split(\" \", expand=True)[1].str.split(\"\\\"\", expand=True)[1]\n",
    "isoforms_df.columns = [\"seqname\", \"insname\", \"feature\", \"start\", \"end\", \"score\", \"strand\", \"score1\", \"info\", \"transcript_id\"]\n",
    "#Only keep DE isoforms\n",
    "brf1_ctrl_inf = pd.read_csv(\"../data/DE_analysis/glaunsinger_Brf1_48hpi_v_Ctrl_48_DE_analysis_mm10.txt\", sep=\"\\s+\")\n",
    "de_transcripts = brf1_ctrl_inf[brf1_ctrl_inf[\"adj.P.Val\"] <= 0.05][\"Transcript\"].unique().tolist()\n",
    "isoforms_df = isoforms_df[isoforms_df[\"transcript_id\"].isin(de_transcripts)]\n",
    "\n",
    "isoforms_df = isoforms_df[[\"seqname\", \"feature\", \"start\", \"end\", \"transcript_id\", \"strand\"]]\n",
    "isoforms_df[\"Score\"] = 0\n",
    "isoforms_df = isoforms_df[[\"seqname\", \"feature\", \"start\", \"end\", \"transcript_id\", \"Score\", \"strand\"]]\n",
    "isoforms_df = isoforms_df[isoforms_df[\"feature\"] == \"transcript\"]\n",
    "isoforms_df = isoforms_df.drop(columns=[\"feature\"])\n",
    "isoforms_df.columns = [\"Chromosome\", \"Transcript_Start\", \"Transcript_End\", \"Name\", \"Score\", \"Strand\"]\n",
    "isoforms_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [],
   "source": [
    "#isoforms_df[\"Chromosome\"].unique().tolist()\n",
    "chroms = ['chr1',\n",
    " 'chr2',\n",
    " 'chr3',\n",
    " 'chr4',\n",
    " 'chr5',\n",
    " 'chr6',\n",
    " 'chr7',\n",
    " 'chr8',\n",
    " 'chr9',\n",
    " 'chr10',\n",
    " 'chr11',\n",
    " 'chr12',\n",
    " 'chr13',\n",
    " 'chr14',\n",
    " 'chr15',\n",
    " 'chr16',\n",
    " 'chr17',\n",
    " 'chr18',\n",
    " 'chr19',\n",
    " 'chrX',\n",
    " 'chrY']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>CDS_Start</th>\n",
       "      <th>CDS_End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Transcript_Start</th>\n",
       "      <th>Transcript_End</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GL456212.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PB.24087.1</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>82189</td>\n",
       "      <td>84538</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>JH584299.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PB.24097.1</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>935049</td>\n",
       "      <td>936476</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>KK082442.1</td>\n",
       "      <td>142719.0</td>\n",
       "      <td>142354.0</td>\n",
       "      <td>PB.24142.42</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>141292</td>\n",
       "      <td>142839</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>KQ030494.1</td>\n",
       "      <td>93892.0</td>\n",
       "      <td>115515.0</td>\n",
       "      <td>PB.24270.2</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>93088</td>\n",
       "      <td>115814</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>150100224.0</td>\n",
       "      <td>150105782.0</td>\n",
       "      <td>PB.1006.28</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>150100100</td>\n",
       "      <td>150105919</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>472</th>\n",
       "      <td>chrX</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PB.23959.3</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>167207095</td>\n",
       "      <td>167209161</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>473</th>\n",
       "      <td>chrY</td>\n",
       "      <td>1010617.0</td>\n",
       "      <td>1016379.0</td>\n",
       "      <td>PB.23977.40</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>1010608</td>\n",
       "      <td>1028591</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>474</th>\n",
       "      <td>chrY</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PB.24050.1</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>75606591</td>\n",
       "      <td>75609221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>475</th>\n",
       "      <td>chrY</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PB.24063.1</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>83165949</td>\n",
       "      <td>83167365</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>476</th>\n",
       "      <td>chrY</td>\n",
       "      <td>87029310.0</td>\n",
       "      <td>87029726.0</td>\n",
       "      <td>PB.24068.1</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>87029087</td>\n",
       "      <td>87030350</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>477 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Chromosome    CDS_Start      CDS_End         Name  Score Strand  \\\n",
       "0    GL456212.1          NaN          NaN   PB.24087.1      0      +   \n",
       "1    JH584299.1          NaN          NaN   PB.24097.1      0      +   \n",
       "2    KK082442.1     142719.0     142354.0  PB.24142.42      0      -   \n",
       "3    KQ030494.1      93892.0     115515.0   PB.24270.2      0      +   \n",
       "4          chr1  150100224.0  150105782.0   PB.1006.28      0      +   \n",
       "..          ...          ...          ...          ...    ...    ...   \n",
       "472        chrX          NaN          NaN   PB.23959.3      0      -   \n",
       "473        chrY    1010617.0    1016379.0  PB.23977.40      0      +   \n",
       "474        chrY          NaN          NaN   PB.24050.1      0      +   \n",
       "475        chrY          NaN          NaN   PB.24063.1      0      +   \n",
       "476        chrY   87029310.0   87029726.0   PB.24068.1      0      +   \n",
       "\n",
       "     Transcript_Start  Transcript_End  \n",
       "0               82189           84538  \n",
       "1              935049          936476  \n",
       "2              141292          142839  \n",
       "3               93088          115814  \n",
       "4           150100100       150105919  \n",
       "..                ...             ...  \n",
       "472         167207095       167209161  \n",
       "473           1010608         1028591  \n",
       "474          75606591        75609221  \n",
       "475          83165949        83167365  \n",
       "476          87029087        87030350  \n",
       "\n",
       "[477 rows x 8 columns]"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "UTR_coords_df.merge(isoforms_df, on=[\"Chromosome\", \"Name\", \"Score\", \"Strand\"], how=\"left\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>CDS_Start</th>\n",
       "      <th>CDS_End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Transcript_Start</th>\n",
       "      <th>Transcript_End</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>KK082442.1</td>\n",
       "      <td>142719.0</td>\n",
       "      <td>142354.0</td>\n",
       "      <td>PB.24142.42</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>141292</td>\n",
       "      <td>142839</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>KQ030494.1</td>\n",
       "      <td>93892.0</td>\n",
       "      <td>115515.0</td>\n",
       "      <td>PB.24270.2</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>93088</td>\n",
       "      <td>115814</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>150100224.0</td>\n",
       "      <td>150105782.0</td>\n",
       "      <td>PB.1006.28</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>150100100</td>\n",
       "      <td>150105919</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>153740469.0</td>\n",
       "      <td>153743855.0</td>\n",
       "      <td>PB.1057.2</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>153740353</td>\n",
       "      <td>153745464</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>153902947.0</td>\n",
       "      <td>153908182.0</td>\n",
       "      <td>PB.1064.20</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>153899946</td>\n",
       "      <td>153908283</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>469</th>\n",
       "      <td>chrX</td>\n",
       "      <td>10718230.0</td>\n",
       "      <td>10718778.0</td>\n",
       "      <td>PB.23410.1</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>10715027</td>\n",
       "      <td>10719693</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>470</th>\n",
       "      <td>chrX</td>\n",
       "      <td>56455488.0</td>\n",
       "      <td>56506988.0</td>\n",
       "      <td>PB.23557.25</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>56454880</td>\n",
       "      <td>56507834</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>471</th>\n",
       "      <td>chrX</td>\n",
       "      <td>140542967.0</td>\n",
       "      <td>140540868.0</td>\n",
       "      <td>PB.23840.13</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>140539537</td>\n",
       "      <td>140543385</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>473</th>\n",
       "      <td>chrY</td>\n",
       "      <td>1010617.0</td>\n",
       "      <td>1016379.0</td>\n",
       "      <td>PB.23977.40</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>1010608</td>\n",
       "      <td>1028591</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>476</th>\n",
       "      <td>chrY</td>\n",
       "      <td>87029310.0</td>\n",
       "      <td>87029726.0</td>\n",
       "      <td>PB.24068.1</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>87029087</td>\n",
       "      <td>87030350</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>430 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Chromosome    CDS_Start      CDS_End         Name  Score Strand  \\\n",
       "2    KK082442.1     142719.0     142354.0  PB.24142.42      0      -   \n",
       "3    KQ030494.1      93892.0     115515.0   PB.24270.2      0      +   \n",
       "4          chr1  150100224.0  150105782.0   PB.1006.28      0      +   \n",
       "5          chr1  153740469.0  153743855.0    PB.1057.2      0      +   \n",
       "6          chr1  153902947.0  153908182.0   PB.1064.20      0      +   \n",
       "..          ...          ...          ...          ...    ...    ...   \n",
       "469        chrX   10718230.0   10718778.0   PB.23410.1      0      +   \n",
       "470        chrX   56455488.0   56506988.0  PB.23557.25      0      +   \n",
       "471        chrX  140542967.0  140540868.0  PB.23840.13      0      -   \n",
       "473        chrY    1010617.0    1016379.0  PB.23977.40      0      +   \n",
       "476        chrY   87029310.0   87029726.0   PB.24068.1      0      +   \n",
       "\n",
       "     Transcript_Start  Transcript_End  \n",
       "2              141292          142839  \n",
       "3               93088          115814  \n",
       "4           150100100       150105919  \n",
       "5           153740353       153745464  \n",
       "6           153899946       153908283  \n",
       "..                ...             ...  \n",
       "469          10715027        10719693  \n",
       "470          56454880        56507834  \n",
       "471         140539537       140543385  \n",
       "473           1010608         1028591  \n",
       "476          87029087        87030350  \n",
       "\n",
       "[430 rows x 8 columns]"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "UTR_coords_df.merge(isoforms_df, on=[\"Chromosome\", \"Name\", \"Score\", \"Strand\"], how=\"left\").dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/hh/vsrq3q5n43zfc6yq4n6bx5n80000gn/T/ipykernel_99052/3728839765.py:7: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  five_prime_pos[\"CDS_Start\"] = five_prime_pos[\"CDS_Start\"] - 1 - 1\n",
      "/var/folders/hh/vsrq3q5n43zfc6yq4n6bx5n80000gn/T/ipykernel_99052/3728839765.py:13: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  five_prime_neg[\"CDS_Start\"] = five_prime_neg[\"CDS_Start\"] - 1 + 1\n",
      "/var/folders/hh/vsrq3q5n43zfc6yq4n6bx5n80000gn/T/ipykernel_99052/3728839765.py:25: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  three_prime_pos[\"CDS_End\"] = three_prime_pos[\"CDS_End\"] - 1 + 1\n",
      "/var/folders/hh/vsrq3q5n43zfc6yq4n6bx5n80000gn/T/ipykernel_99052/3728839765.py:31: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  three_prime_neg[\"CDS_End\"] = three_prime_neg[\"CDS_End\"] - 1 - 1\n"
     ]
    }
   ],
   "source": [
    "UTR_coords_df = UTR_coords_df.merge(isoforms_df, on=[\"Chromosome\", \"Name\", \"Score\", \"Strand\"], how=\"left\").dropna()\n",
    "\n",
    "#5' UTR\n",
    "five_prime_UTR_df = UTR_coords_df[[\"Chromosome\", \"Transcript_Start\", \"Transcript_End\", \"CDS_Start\", \"Name\", \"Score\", \"Strand\"]]\n",
    "five_prime_pos = five_prime_UTR_df[five_prime_UTR_df[\"Strand\"] == \"+\"]\n",
    "#Make zero-based first and then move one base pair away from CDS start\n",
    "five_prime_pos[\"CDS_Start\"] = five_prime_pos[\"CDS_Start\"] - 1 - 1\n",
    "five_prime_pos = five_prime_pos.drop(columns=[\"Transcript_End\"])\n",
    "five_prime_pos[\"Transcript_Start\"] = five_prime_pos[\"Transcript_Start\"] - 1\n",
    "five_prime_pos.columns = [\"Chromosome\", \"Start\", \"End\", \"Name\", \"Score\", \"Strand\"]\n",
    "\n",
    "five_prime_neg = five_prime_UTR_df[five_prime_UTR_df[\"Strand\"] == \"-\"]\n",
    "five_prime_neg[\"CDS_Start\"] = five_prime_neg[\"CDS_Start\"] - 1 + 1\n",
    "five_prime_neg = five_prime_neg.drop(columns=[\"Transcript_Start\"])\n",
    "five_prime_neg = five_prime_neg[[\"Chromosome\", \"CDS_Start\", \"Transcript_End\", \"Name\", \"Score\", \"Strand\"]]\n",
    "five_prime_neg[\"Transcript_End\"] = five_prime_neg[\"Transcript_End\"] - 1\n",
    "five_prime_neg.columns = [\"Chromosome\", \"Start\", \"End\", \"Name\", \"Score\", \"Strand\"]\n",
    "five_prime_UTR_df = pd.concat([five_prime_pos, five_prime_neg])\n",
    "five_prime_UTR_df = five_prime_UTR_df[five_prime_UTR_df['Chromosome'].isin(chroms)]\n",
    "five_prime_UTR_df = five_prime_UTR_df.drop(five_prime_UTR_df[five_prime_UTR_df.End < five_prime_UTR_df.Start].index)\n",
    "\n",
    "#3' UTR\n",
    "three_prime_UTR_df = UTR_coords_df[[\"Chromosome\", \"CDS_End\", \"Transcript_Start\", \"Transcript_End\", \"Name\", \"Score\", \"Strand\"]]\n",
    "three_prime_pos = three_prime_UTR_df[three_prime_UTR_df[\"Strand\"] == \"+\"]\n",
    "three_prime_pos[\"CDS_End\"] = three_prime_pos[\"CDS_End\"] - 1 + 1\n",
    "three_prime_pos = three_prime_pos.drop(columns=[\"Transcript_Start\"])\n",
    "three_prime_pos[\"Transcript_End\"] = three_prime_pos[\"Transcript_End\"] - 1\n",
    "three_prime_pos.columns = [\"Chromosome\", \"Start\", \"End\", \"Name\", \"Score\", \"Strand\"]\n",
    "\n",
    "three_prime_neg = three_prime_UTR_df[three_prime_UTR_df[\"Strand\"] == \"-\"]\n",
    "three_prime_neg[\"CDS_End\"] = three_prime_neg[\"CDS_End\"] - 1 - 1\n",
    "three_prime_neg = three_prime_neg.drop(columns=[\"Transcript_End\"])\n",
    "three_prime_neg = three_prime_neg[[\"Chromosome\", \"Transcript_Start\", \"CDS_End\", \"Name\", \"Score\", \"Strand\"]]\n",
    "three_prime_neg[\"Transcript_Start\"] = three_prime_neg[\"Transcript_Start\"] - 1\n",
    "three_prime_neg.columns = [\"Chromosome\", \"Start\", \"End\", \"Name\", \"Score\", \"Strand\"]\n",
    "three_prime_UTR_df = pd.concat([three_prime_pos, three_prime_neg])\n",
    "three_prime_UTR_df = three_prime_UTR_df[three_prime_UTR_df['Chromosome'].isin(chroms)]\n",
    "three_prime_UTR_df = three_prime_UTR_df.drop(three_prime_UTR_df[three_prime_UTR_df.End < three_prime_UTR_df.Start].index)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>150100099.0</td>\n",
       "      <td>150100222.0</td>\n",
       "      <td>PB.1006.28</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>153740352.0</td>\n",
       "      <td>153740467.0</td>\n",
       "      <td>PB.1057.2</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>153899945.0</td>\n",
       "      <td>153902945.0</td>\n",
       "      <td>PB.1064.20</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr1</td>\n",
       "      <td>171329617.0</td>\n",
       "      <td>171338314.0</td>\n",
       "      <td>PB.1242.16</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>chr1</td>\n",
       "      <td>91053478.0</td>\n",
       "      <td>91053656.0</td>\n",
       "      <td>PB.654.62</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>464</th>\n",
       "      <td>chr9</td>\n",
       "      <td>111056428.0</td>\n",
       "      <td>111057427.0</td>\n",
       "      <td>PB.12944.27</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>466</th>\n",
       "      <td>chrX</td>\n",
       "      <td>12061607.0</td>\n",
       "      <td>12080552.0</td>\n",
       "      <td>PB.23409.12</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>467</th>\n",
       "      <td>chrX</td>\n",
       "      <td>12061607.0</td>\n",
       "      <td>12080552.0</td>\n",
       "      <td>PB.23409.34</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>468</th>\n",
       "      <td>chrX</td>\n",
       "      <td>12061607.0</td>\n",
       "      <td>12081267.0</td>\n",
       "      <td>PB.23409.39</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>471</th>\n",
       "      <td>chrX</td>\n",
       "      <td>140542967.0</td>\n",
       "      <td>140543384.0</td>\n",
       "      <td>PB.23840.13</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>424 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Chromosome        Start          End         Name  Score Strand\n",
       "4         chr1  150100099.0  150100222.0   PB.1006.28      0      +\n",
       "5         chr1  153740352.0  153740467.0    PB.1057.2      0      +\n",
       "6         chr1  153899945.0  153902945.0   PB.1064.20      0      +\n",
       "10        chr1  171329617.0  171338314.0   PB.1242.16      0      +\n",
       "18        chr1   91053478.0   91053656.0    PB.654.62      0      +\n",
       "..         ...          ...          ...          ...    ...    ...\n",
       "464       chr9  111056428.0  111057427.0  PB.12944.27      0      -\n",
       "466       chrX   12061607.0   12080552.0  PB.23409.12      0      -\n",
       "467       chrX   12061607.0   12080552.0  PB.23409.34      0      -\n",
       "468       chrX   12061607.0   12081267.0  PB.23409.39      0      -\n",
       "471       chrX  140542967.0  140543384.0  PB.23840.13      0      -\n",
       "\n",
       "[424 rows x 6 columns]"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "five_prime_UTR_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>150105782.0</td>\n",
       "      <td>150105918.0</td>\n",
       "      <td>PB.1006.28</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>153743855.0</td>\n",
       "      <td>153745463.0</td>\n",
       "      <td>PB.1057.2</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>153908182.0</td>\n",
       "      <td>153908282.0</td>\n",
       "      <td>PB.1064.20</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr1</td>\n",
       "      <td>171341204.0</td>\n",
       "      <td>171341681.0</td>\n",
       "      <td>PB.1242.16</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>chr1</td>\n",
       "      <td>91112377.0</td>\n",
       "      <td>91115529.0</td>\n",
       "      <td>PB.654.62</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>464</th>\n",
       "      <td>chr9</td>\n",
       "      <td>111055177.0</td>\n",
       "      <td>111055344.0</td>\n",
       "      <td>PB.12944.27</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>466</th>\n",
       "      <td>chrX</td>\n",
       "      <td>12036742.0</td>\n",
       "      <td>12037624.0</td>\n",
       "      <td>PB.23409.12</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>467</th>\n",
       "      <td>chrX</td>\n",
       "      <td>12037325.0</td>\n",
       "      <td>12037624.0</td>\n",
       "      <td>PB.23409.34</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>468</th>\n",
       "      <td>chrX</td>\n",
       "      <td>12037325.0</td>\n",
       "      <td>12037624.0</td>\n",
       "      <td>PB.23409.39</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>471</th>\n",
       "      <td>chrX</td>\n",
       "      <td>140539536.0</td>\n",
       "      <td>140540866.0</td>\n",
       "      <td>PB.23840.13</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>428 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Chromosome        Start          End         Name  Score Strand\n",
       "4         chr1  150105782.0  150105918.0   PB.1006.28      0      +\n",
       "5         chr1  153743855.0  153745463.0    PB.1057.2      0      +\n",
       "6         chr1  153908182.0  153908282.0   PB.1064.20      0      +\n",
       "10        chr1  171341204.0  171341681.0   PB.1242.16      0      +\n",
       "18        chr1   91112377.0   91115529.0    PB.654.62      0      +\n",
       "..         ...          ...          ...          ...    ...    ...\n",
       "464       chr9  111055177.0  111055344.0  PB.12944.27      0      -\n",
       "466       chrX   12036742.0   12037624.0  PB.23409.12      0      -\n",
       "467       chrX   12037325.0   12037624.0  PB.23409.34      0      -\n",
       "468       chrX   12037325.0   12037624.0  PB.23409.39      0      -\n",
       "471       chrX  140539536.0  140540866.0  PB.23840.13      0      -\n",
       "\n",
       "[428 rows x 6 columns]"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "three_prime_UTR_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [],
   "source": [
    "bf = pybed.BedFrame.from_frame(meta=[], data=five_prime_UTR_df)\n",
    "bf.to_file('../data/bed_files/DE_isoforms_5_prime_coords_ctrl48_brf148.bed')\n",
    "\n",
    "bf = pybed.BedFrame.from_frame(meta=[], data=three_prime_UTR_df)\n",
    "bf.to_file('../data/bed_files/DE_isoforms_3_prime_coords_ctrl48_brf148.bed')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Transcript_Start</th>\n",
       "      <th>Transcript_End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>50322</th>\n",
       "      <td>chr1</td>\n",
       "      <td>55003175</td>\n",
       "      <td>55027395</td>\n",
       "      <td>PB.316.57</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75210</th>\n",
       "      <td>chr1</td>\n",
       "      <td>74288252</td>\n",
       "      <td>74304619</td>\n",
       "      <td>PB.468.7</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94407</th>\n",
       "      <td>chr1</td>\n",
       "      <td>84034264</td>\n",
       "      <td>84284599</td>\n",
       "      <td>PB.558.4</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126881</th>\n",
       "      <td>chr1</td>\n",
       "      <td>91053479</td>\n",
       "      <td>91115530</td>\n",
       "      <td>PB.654.62</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128768</th>\n",
       "      <td>chr1</td>\n",
       "      <td>91453766</td>\n",
       "      <td>91459301</td>\n",
       "      <td>PB.666.13</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4455885</th>\n",
       "      <td>chrY</td>\n",
       "      <td>87029087</td>\n",
       "      <td>87030350</td>\n",
       "      <td>PB.24068.1</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4456545</th>\n",
       "      <td>GL456212.1</td>\n",
       "      <td>82189</td>\n",
       "      <td>84538</td>\n",
       "      <td>PB.24087.1</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4456954</th>\n",
       "      <td>JH584299.1</td>\n",
       "      <td>935049</td>\n",
       "      <td>936476</td>\n",
       "      <td>PB.24097.1</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4462451</th>\n",
       "      <td>KK082442.1</td>\n",
       "      <td>141292</td>\n",
       "      <td>142839</td>\n",
       "      <td>PB.24142.42</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4474510</th>\n",
       "      <td>KQ030494.1</td>\n",
       "      <td>93088</td>\n",
       "      <td>115814</td>\n",
       "      <td>PB.24270.2</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>477 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         Chromosome  Transcript_Start  Transcript_End         Name  Score  \\\n",
       "50322          chr1          55003175        55027395    PB.316.57      0   \n",
       "75210          chr1          74288252        74304619     PB.468.7      0   \n",
       "94407          chr1          84034264        84284599     PB.558.4      0   \n",
       "126881         chr1          91053479        91115530    PB.654.62      0   \n",
       "128768         chr1          91453766        91459301    PB.666.13      0   \n",
       "...             ...               ...             ...          ...    ...   \n",
       "4455885        chrY          87029087        87030350   PB.24068.1      0   \n",
       "4456545  GL456212.1             82189           84538   PB.24087.1      0   \n",
       "4456954  JH584299.1            935049          936476   PB.24097.1      0   \n",
       "4462451  KK082442.1            141292          142839  PB.24142.42      0   \n",
       "4474510  KQ030494.1             93088          115814   PB.24270.2      0   \n",
       "\n",
       "        Strand  \n",
       "50322        -  \n",
       "75210        -  \n",
       "94407        -  \n",
       "126881       +  \n",
       "128768       -  \n",
       "...        ...  \n",
       "4455885      +  \n",
       "4456545      +  \n",
       "4456954      +  \n",
       "4462451      -  \n",
       "4474510      +  \n",
       "\n",
       "[477 rows x 6 columns]"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "isoforms_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Extract promoter coordinates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/hh/vsrq3q5n43zfc6yq4n6bx5n80000gn/T/ipykernel_99052/686762972.py:4: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  promoter_pos[\"Transcript_Start\"] = promoter_pos[\"Transcript_Start\"] - 1\n",
      "/var/folders/hh/vsrq3q5n43zfc6yq4n6bx5n80000gn/T/ipykernel_99052/686762972.py:5: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  promoter_pos[\"Start\"] = promoter_pos[\"Transcript_Start\"] - 4000\n",
      "/var/folders/hh/vsrq3q5n43zfc6yq4n6bx5n80000gn/T/ipykernel_99052/686762972.py:10: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  promoter_neg[\"Transcript_End\"] = promoter_neg[\"Transcript_End\"] - 1\n",
      "/var/folders/hh/vsrq3q5n43zfc6yq4n6bx5n80000gn/T/ipykernel_99052/686762972.py:11: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  promoter_neg[\"End\"] = promoter_neg[\"Transcript_End\"] + 4000\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>126881</th>\n",
       "      <td>chr1</td>\n",
       "      <td>91049478</td>\n",
       "      <td>91053478</td>\n",
       "      <td>PB.654.62</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145587</th>\n",
       "      <td>chr1</td>\n",
       "      <td>107586031</td>\n",
       "      <td>107590031</td>\n",
       "      <td>PB.756.22</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>192874</th>\n",
       "      <td>chr1</td>\n",
       "      <td>150096099</td>\n",
       "      <td>150100099</td>\n",
       "      <td>PB.1006.28</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>207160</th>\n",
       "      <td>chr1</td>\n",
       "      <td>153736352</td>\n",
       "      <td>153740352</td>\n",
       "      <td>PB.1057.2</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>208273</th>\n",
       "      <td>chr1</td>\n",
       "      <td>153895945</td>\n",
       "      <td>153899945</td>\n",
       "      <td>PB.1064.20</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4322929</th>\n",
       "      <td>chrX</td>\n",
       "      <td>12080552</td>\n",
       "      <td>12084552</td>\n",
       "      <td>PB.23409.12</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4323372</th>\n",
       "      <td>chrX</td>\n",
       "      <td>12080552</td>\n",
       "      <td>12084552</td>\n",
       "      <td>PB.23409.34</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4323448</th>\n",
       "      <td>chrX</td>\n",
       "      <td>12081267</td>\n",
       "      <td>12085267</td>\n",
       "      <td>PB.23409.39</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4419858</th>\n",
       "      <td>chrX</td>\n",
       "      <td>140543384</td>\n",
       "      <td>140547384</td>\n",
       "      <td>PB.23840.13</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4447738</th>\n",
       "      <td>chrX</td>\n",
       "      <td>167209160</td>\n",
       "      <td>167213160</td>\n",
       "      <td>PB.23959.3</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>473 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chromosome      Start        End         Name  Score Strand\n",
       "126881        chr1   91049478   91053478    PB.654.62      0      +\n",
       "145587        chr1  107586031  107590031    PB.756.22      0      +\n",
       "192874        chr1  150096099  150100099   PB.1006.28      0      +\n",
       "207160        chr1  153736352  153740352    PB.1057.2      0      +\n",
       "208273        chr1  153895945  153899945   PB.1064.20      0      +\n",
       "...            ...        ...        ...          ...    ...    ...\n",
       "4322929       chrX   12080552   12084552  PB.23409.12      0      -\n",
       "4323372       chrX   12080552   12084552  PB.23409.34      0      -\n",
       "4323448       chrX   12081267   12085267  PB.23409.39      0      -\n",
       "4419858       chrX  140543384  140547384  PB.23840.13      0      -\n",
       "4447738       chrX  167209160  167213160   PB.23959.3      0      -\n",
       "\n",
       "[473 rows x 6 columns]"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "promoter_df = isoforms_df[[\"Chromosome\", \"Transcript_Start\", \"Transcript_End\", \"Name\", \"Score\", \"Strand\"]]\n",
    "\n",
    "promoter_pos = promoter_df[promoter_df[\"Strand\"] == \"+\"]\n",
    "promoter_pos[\"Transcript_Start\"] = promoter_pos[\"Transcript_Start\"] - 1\n",
    "promoter_pos[\"Start\"] = promoter_pos[\"Transcript_Start\"] - 4000\n",
    "promoter_pos = promoter_pos[[\"Chromosome\", \"Start\", \"Transcript_Start\", \"Name\", \"Score\", \"Strand\"]]\n",
    "promoter_pos.columns = [\"Chromosome\", \"Start\", \"End\", \"Name\", \"Score\", \"Strand\"]\n",
    "\n",
    "promoter_neg = promoter_df[promoter_df[\"Strand\"] == \"-\"]\n",
    "promoter_neg[\"Transcript_End\"] = promoter_neg[\"Transcript_End\"] - 1\n",
    "promoter_neg[\"End\"] = promoter_neg[\"Transcript_End\"] + 4000\n",
    "promoter_neg = promoter_neg[[\"Chromosome\", \"Transcript_End\", \"End\", \"Name\", \"Score\", \"Strand\"]]\n",
    "promoter_neg.columns = [\"Chromosome\", \"Start\", \"End\", \"Name\", \"Score\", \"Strand\"]\n",
    "\n",
    "promoter_df = pd.concat([promoter_pos, promoter_neg])\n",
    "promoter_df = promoter_df[promoter_df['Chromosome'].isin(chroms)]\n",
    "promoter_df = promoter_df.drop(promoter_df[promoter_df.End < promoter_df.Start].index)\n",
    "promoter_df\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [],
   "source": [
    "bf = pybed.BedFrame.from_frame(meta=[], data=promoter_df)\n",
    "bf.to_file('../data/bed_files/DE_isoforms_promoter_up4kb_coords_ctrl48_brf148.bed')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
